src/bin/pg_autoctl/cli_create_drop_node.c

```diff
@@ -614,73 +720,171 @@ cli_drop_node(int argc, char **argv)
+				if (!keeper_config_read_file(&config,
+											 missingPgdataIsOk,
+											 pgIsNotRunningIsOk,
+											 monitorDisabledIsOk))
+				{
+					exit(EXIT_CODE_BAD_CONFIG);
+				}
+				break;
+			}
+
+			default:
+			{
+				log_fatal("Unrecognized configuration file \"%s\"",
+						  config.pathnames.config);
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+		}
 
-	if (directory_exists(config.pgSetup.pgdata))
+		log_trace("Found pg_ctl at \"%s\" in config file \"%s\"",
+				  config.pgSetup.pg_ctl,
+				  config.pathnames.config);
+	}
+	else
 	{
-		log_warn("Postgres Data Directory \"%s\" has been preserved",
-				 config.pgSetup.pgdata);
+		/* all we really need now is pg_ctl */
+		set_first_pgctl(&(config.pgSetup));
 	}
 
-	log_info("drop node keeps your data and setup safe, you can still run "
-			 "Postgres or re-join the pg_auto_failover cluster later");
-	log_info("HINT: to completely remove your local Postgres instance and "
-			 "setup, consider the command \"pg_autoctl destroy\"");
-
 	/*
-	 * Now also stop Postgres and the keeper itself, keeper first.
-	 *
-	 * We need to stop Postgres because otherwise we won't be able to drop the
-	 * replication slot on the other node, because it's still active.
+	 * Now also stop the pg_autoctl process.
 	 */
-	if (read_pidfile(keeper.config.pathnames.pid, &pid))
+	if (file_exists(config.pathnames.pid))
 	{
-		if (kill(pid, SIGTERM) != 0)
+		pid_t pid = 0;
+
+		if (read_pidfile(config.pathnames.pid, &pid))
 		{
-			log_error("Failed to send %s to the keeper's pid %d: %s",
-					  strsignal(SIGTERM), pid, strerror(errno));
-			exit(EXIT_CODE_INTERNAL_ERROR);
+			log_info("An instance of this keeper is running with PID %d, "
+					 "stopping it.", pid);
+
+			if (kill(pid, SIGQUIT) != 0)
+			{
+				log_error("Failed to send SIGQUIT to the keeper's pid %d: %s",
+						  pid, strerror(errno));
+				exit(EXIT_CODE_INTERNAL_ERROR);
+			}
 		}
-		else
+	}
+
+	/* only keeper_remove when we still have a state file around */
+	if (file_exists(config.pathnames.state))
+	{
+		bool ignoreMonitorErrors = true;
+
+		/* keeper_remove uses log_info() to explain what's happening */
+		if (!keeper_remove(&keeper, &config, ignoreMonitorErrors))
 		{
-			log_info("Stopped the pg_autoctl service.");
+			log_fatal("Failed to remove local node from the pg_auto_failover "
+					  "monitor, see above for details");
+
+			exit(EXIT_CODE_BAD_STATE);
 		}
+
+		log_info("Removed pg_autoctl node at \"%s\" from the monitor and "
+				 "removed the state file \"%s\"",
+				 config.pgSetup.pgdata,
+				 config.pathnames.state);
+	}
+	else
+	{
+		log_warn("Skipping node removal from the monitor: "
+				 "state file \"%s\" does not exist",
+				 config.pathnames.state);
 	}
 
-	if (pg_ctl_stop(config.pgSetup.pg_ctl, config.pgSetup.pgdata))
+	/*
+	 * Either --destroy the whole Postgres cluster and configuraiton, or leave
+	 * enough behind us that it's possible to re-join a formation later.
+	 */
+	if (dropAndDestroy)
```

Comment:
destroy also includes drop right ? why don't be make this earlier in the function ? This way we might need less changes in the drop implementation

Reply:
Well the rest of the actions are common to both `drop node` and `drop node --destroy`, so we don't need the test until way down in the implementation. The new code is reading the configuration file to grab the `postgresql.pg_ctl` parameter and make sure that's the one we use, where before we would just use `set_first_pgctl` and might use `pg_ctl` from Postgres 12 on a Postgres 11 cluster.